### PR TITLE
Router-controlled page state.

### PR DESCRIPTION
**Pull request by nerdstrike**
Not happy with the code, but it seems to work as desired. Tests completely ignored

### CODE REVIEW

File: frontend/src/views/RunView.vue [hunk 1]
```diff
@@ -32,6 +35,37 @@ let activeWell = reactive({
...
 getUserName((email) => { user.value = email }).then();
 
+watch(() => route.query, (after, before) => {
+  console.log(before)
+  if (
+    (after.label || after.run)
+    && (
+      before == undefined
+      || (
+        before.label && before.run && after.label != before.label && after.run != before.run
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:45]: It is conceavable for teh value of run not to change. Therefore, the condition should be
`&& (after.label != before.label || after.run != before.run)`

File: frontend/src/views/RunView.vue [hunk 2]
```diff
@@ -32,6 +35,37 @@ let activeWell = reactive({
 let user = ref(null);
 getUserName((email) => { user.value = email }).then();
 
+watch(() => route.query, (after, before) => {
+  console.log(before)
+  if (
+    (after.label || after.run)
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:41]: Probably both should be defined

File: frontend/src/views/RunView.vue [hunk 3]
```diff
@@ -32,6 +35,37 @@ let activeWell = reactive({
...
+    console.log('Noticed change in run/label')
+    // Somehow we need to capture the other parameter in case both have not been set
+    loadWellDetail(after.run, after.label)
+  }
+  if (after.activeTab && (before === undefined || after.activeTab != before.activeTab)) {
+    changeTab(after.activeTab)
+  }
+  if (after.page && (before == undefined || after.page != before.page)) {
+    changePage(after.page)
+  }
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:58]: We suspect that we will have two trips to the server and render twice.

File: frontend/src/views/RunView.vue [hunk 4]
```diff
@@ -80,12 +114,22 @@ function loadWells(status, page, pageSize) {
 function changeTab(selectedTab) {
   // To be triggered from Tab elements to load different data sets
   // Reset page to 1 on tab change
+  console.log("Changing tab to " + selectedTab)
   activePage.value = 1;
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:118]: Should be doing this, should be getting info about the requested page from URL.

File: frontend/src/views/RunView.vue [hunk 5]
```diff
@@ -94,13 +155,37 @@ function externalTabChange(tabName) {
...
 
+function updateUrlQuery(newParams) {
+  // Merges the current URL query with new parameters
+  // The calling code does not need to replicate all properties from the
+  // previous URL query.
+  // route.query is read-only. We must copy it before modifying
+  let existingSettings = Object.assign({}, route.query)
+  let changed = false
+  for (let k in newParams) {
+    if (existingSettings[k] && existingSettings[k] != newParams[k] || !existingSettings[k]) {
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:166]: `if (!existingSettings[k] || existingSettings[k] != newParams[k] )` ?

File: frontend/src/views/RunView.vue [hunk 6]
```diff
@@ -32,6 +35,55 @@ let activeWell = reactive({
...
+    )
+  ) {
+    // Somehow we need to capture the other parameter in case both have not been set
+    loadWellDetail(after.qcRun, after.qcLabel)
+  }
+
+  // Handle changes of tab in the table of wells.
+  // Changes to selected tab negates a page change operation
+  if (after.activeTab && (before === undefined || after.activeTab != before.activeTab)) {
+    changeTab(after.activeTab, newPage)
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:72]: Call `loadWells` directly?

File: frontend/src/views/RunView.vue [hunk 7]
```diff
@@ -32,6 +35,55 @@ let activeWell = reactive({
...
+    activeTab.value = after.activeTab;
+    if (before == undefined || after.page != before.page) {
+      activePage.value = newPage;
+    }
+  } else if (after.page && (before == undefined || after.page != before.page)) {
+    changePage(newPage)
```
Comment (mgcam) [inline frontend/src/views/RunView.vue:78]: Call `loadWells` directly?

File: frontend/src/views/RunView.vue [hunk 8]
```diff
@@ -94,13 +155,37 @@ function externalTabChange(tabName) {
   activeTab.value = tabName;
 }
 
+function updateUrlQuery(newParams) {
```
[the remark this answers is not in the document]
Comment (nerdstrike) [reply frontend/src/views/RunView.vue:158]: Agree. I will try to split it off into its own library. Not sure how the router will take to being tested without Vue but we'll see

[the remark this answers is not in the document]
Comment (nerdstrike) [reply frontend/src/views/RunView.vue:159]: Got a very vulgar solution for testing.

**Pull request merged**